src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -275,7 +291,7 @@ bool MergeTreeDataMergerMutator::selectPartsToMerge(
         prev_part = &part;
```

Comment:
The bug is here. If we found a new partition we will reset `prev_part` to `nullptr` at line 255, but after that restore it back. So on the next iteration, we will skip `can_merge_callback` check for the first part in the partition, because `partition_id != *prev_partition_id` and unconditionally add this part to the candidates list. Even if this part already participates in merge/mutation. MergeSelectors can easily select this part and assign a merge.
Previously we have seen very similar behavior for ReplicatedMergeTree in 00992 and 00993 tests.

Reply:
By the way, does it lead only to extra merges being assigned (with no harm) + subsequent exception with logical error in code? Or something more dangerous?

Reply:
I think the worst consequence is unprocessable entries https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/StorageReplicatedMergeTree.cpp#L1465 in the ReplicatedMergeTreeQueue (we have seen this rare error in 00992 and 00993 tests).